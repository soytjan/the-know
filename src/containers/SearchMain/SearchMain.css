@import url("https://fonts.googleapis.com/css?family=Open+Sans:300,400|Poiret+One");
.SearchMain input {
  font-size: 1.2em;
  padding: 5px;
  color: #383838;
  border: 2px solid #ECF0F1; }

.SearchMain button {
  font-size: 1em;
  padding: 5px;
  width: 120px; }

.SearchMain .input-search {
  width: 358px; }

.SearchMain .input-location {
  width: 150px;
  text-transform: capitalize; }
