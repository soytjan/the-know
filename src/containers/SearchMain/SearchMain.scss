@import "../../variables.scss";

.SearchMain {
  input {
    font-size: 1.2em;
    padding: 5px;
    color: $black;
    border: 2px solid $input-border-color;
  }

  button {
    font-size: 1em;
    border: 2px solid $search-btn-border-color;
    background-color: $white;
    color: $search-btn-font;
    padding: 5px;
    width: 120px;
  }

  .input-search {
    width: 358px;
  }

  .input-location {
    width: 150px;
    text-transform: capitalize;
  }
}