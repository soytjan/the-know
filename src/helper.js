import keys from './api/keys.js';
const corsAnywhereUrl = 'https://cors-anywhere.herokuapp.com/';

// if this gets really big think about organize api calls into their own api files
export const getGeoLocation = async () => {
  try {
    const url = `https://www.googleapis.com/geolocation/v1/geolocate?key=${keys.googleMapsApiKey}`
    const response = await fetch(url, {
      method: 'POST',
      header: { 'content-type': 'application/json' }
     })

    return response;
  } catch (error) {
    throw (error);
  }
}

// export const initialFetchWithCoords = async (coords) => {
//   try {
//     const url = `${corsAnywhereUrl}http://api.eventful.com/json/events/search?...&where=${coords.lat},${coords.lng}&within=25&&app_key=${keys.eventfulKey}`;
//     const response = await fetch(url, {mode: 'cors'});

//     if (response.status > 226) {
//       throw new Error('could not get city event data');
//       // handleError function that handles error
//     } else {
//       const responseJson = await response.json();
//       return responseJson
//     }
//   } catch (error) {
//     throw (error);
//   }
// };

export const fetchCityData = async (location) => {
  const coords = location.coordinates;
  
  try {
    const url = `${corsAnywhereUrl}http://api.eventful.com/json/events/search?...&where=${coords.lat},${coords.lng}&within=25&&app_key=${keys.eventfulKey}`;
    const response = await fetch(url, {mode: 'cors'});

    if (response.status > 226) {
      throw new Error('could not get city event data');
      // handleError function that handles error
    } else {
      const responseJson = await response.json();
      return responseJson
    }
  } catch (error) {
    throw (error)
  }
}

export const getAddressCoords = async (location) => {
  try {
    const url = `https://maps.googleapis.com/maps/api/geocode/json?address=${location}&key=${keys.googleMapsApiKey}`
    const response = await fetch(url);
    if (response.status > 226) {
      throw new Error('could not get address coordinates');
    } else {
      const responseJson = await response.json();
      return responseJson
    }
  } catch (error) {
    throw (error)
  }
}

export const cleanAddressCoords = (address) => {
  const geocode = address.results[0];
  const cleanLocation = {
    address: geocode.formatted_address,
    coordinates: geocode.geometry.location,
  }

  return cleanLocation;
}  


// minimize some of the repetition
// handleError and pass in error as an argument
const genApiUrl = (type, location) => {
  const coords = location.coordinates;
  
  if (!coords) {
    // need to figure out a better way to handle this
    // default returns url for denver
    // look into pulling in the localstorage  or current location address 
    return `${corsAnywhereUrl}http://api.eventful.com/json/events/search?...&location=Denver&date=Future&app_key=${keys.eventfulKey}`
  }

  let category; 
  switch(type) {
    case 'music':
      category = '&category=music';
      break;
    case 'food':
      category = '&category=food';
      break;
    case 'culture':
      category = '&category=attractions';
      break;
    case 'nightlife':
      category = '&category=singles_social';
      break;
    default: 
      category = '';
  }

  return `${corsAnywhereUrl}http://api.eventful.com/json/events/search?......&where=${coords.lat},${coords.lng}&within=25&&app_key=${keys.eventfulKey}${category}`
}

export const getCategoryData = async (type, location) => {
  try {
    const url = genApiUrl(type, location);
    const response = await fetch(url, {mode: 'cors'});

    if (response.status > 226) {
      throw new Error('could not get city event data');
      // handleError function that handles error
    } else {
      const responseJson = await response.json();
      return responseJson
    }
  } catch (error) {
    throw (error)
  }
}

export const cleanEventData = (cityData) => {
  const events = cityData.events.event;

  const cleanedEvents = events.map(event => {
    return {
      title: event.title,
      description: event.description,
      venueName: event.venue_name,
      region: event.region_abbr,
      postalCode: event.postal_code,
      city: event.city_name,
      startTime: event.start_time,
      venueAddress: event.venue_address,
      image: event.image,
      url: event.url,
      id: event.id,
    }
  })

  return cleanedEvents;
}