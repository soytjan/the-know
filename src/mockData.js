export const mockEventData = {
  "last_item": null,
  "total_items": "9621",
  "first_item": null,
  "page_number": "1",
  "page_size": "10",
  "page_items": null,
  "search_time": "0.049",
  "page_count": "963",
  "events": {
    "event": [
      {
        "watching_count": null,
        "olson_path": "America/Denver",
        "calendar_count": null,
        "comment_count": null,
        "region_abbr": "CO",
        "postal_code": null,
        "going_count": null,
        "all_day": "0",
        "latitude": "39.7392358",
        "groups": null,
        "url": "http://denver.eventful.com/events/thursday-work-day-july-5-multiple-neighborhoods-/E0-001-111209195-8?utm_source=apis&utm_medium=apim&utm_campaign=apic",
        "id": "E0-001-111209195-8",
        "privacy": "1",
        "city_name": "Denver",
        "link_count": null,
        "longitude": "-104.990251",
        "country_name": "United States",
        "country_abbr": "USA",
        "region_name": "Colorado",
        "start_time": "2018-07-05 09:00:00",
        "tz_id": null,
        "description": " <strong>Sponsored by:</strong> Extreme Community Makeover<br> <strong>Minimum age:</strong> 14<br> Extreme Community Makeover (ECM) is designed to invite volunteers into a partnership with residents of specific Denver neighborhoods to complete home and neighborhood improvement projects. These projects include graffiti removal, landscaping, painting, yard work, or other projects identified by the . . .",
        "modified": "2018-02-19 06:12:10",
        "venue_display": "0",
        "tz_country": null,
        "performers": null,
        "title": "Thursday Work Day, July 5 - Multiple Neighborhoods",
        "venue_address": "Denver, Colorado, United States",
        "geocode_type": "EVDB Geocoder",
        "tz_olson_path": null,
        "recur_string": null,
        "calendars": null,
        "owner": "evdb",
        "going": null,
        "country_abbr2": "US",
        "image": null,
        "created": "2018-02-01 04:56:32",
        "venue_id": "V0-001-000179395-7",
        "tz_city": null,
        "stop_time": null,
        "venue_name": "Denver, Colorado, United States",
        "venue_url": "http://denver.eventful.com/venues/denver-colorado-united-states-/V0-001-000179395-7?utm_source=apis&utm_medium=apim&utm_campaign=apic"
      }
    ]
  }
}

export const mockCleanEventData = [
  {
    title: "Thursday Work Day, July 5 - Multiple Neighborhoods",
    description: " <strong>Sponsored by:</strong> Extreme Community Makeover<br> <strong>Minimum age:</strong> 14<br> Extreme Community Makeover (ECM) is designed to invite volunteers into a partnership with residents of specific Denver neighborhoods to complete home and neighborhood improvement projects. These projects include graffiti removal, landscaping, painting, yard work, or other projects identified by the . . .",
    category: "event",
    venueName: "Denver, Colorado, United States", 
    region: "CO",
    postalCode: null,
    city: "Denver",
    startTime: "2018-07-05 09:00:00",
    venueAddress: "Denver, Colorado, United States",
    image: null,
    url: "http://denver.eventful.com/events/thursday-work-day-july-5-multiple-neighborhoods-/E0-001-111209195-8?utm_source=apis&utm_medium=apim&utm_campaign=apic",
    id: "E0-001-111209195-8"
  }
]