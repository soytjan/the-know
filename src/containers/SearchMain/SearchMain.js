import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import './SearchMain.css';


export class SearchMain extends Component {
  constructor(props) {
    super(props);
    this.state = {
      eventSearch: '',
      location: this.props.location
    }
  }

  handleChange = (e) => {
    const { value, name } = e.target;

    this.setState({ [name]: value });
  }

  handleSubmit = () => {
    // will need to make another API call and get back information
    // clean information from API
    // clear out storage
    // then put in something new
    // change the location in store
    // change the location in localStorage
  }

  render() {
    return (
      <div className='SearchMain'>
        <input
          onChange={this.handleChange}
          name='eventSearch'
          value={this.state.eventSearch}
          placeholder='Find events in your area'
          className='input-search' 
          type="text"
        />
        <input 
          onChange={this.handleChange}
          name='location'
          value={this.state.location}
          placeholder='Where?' 
          className='input-location'
          type="text"
        />
        <button>SEARCH</button>
      </div>
    )
  }
}

SearchMain.propTypes = {
  location: PropTypes.string,
};

const mapStateToProps = state => ({
  location: state.location
})

export default connect(mapStateToProps)(SearchMain);