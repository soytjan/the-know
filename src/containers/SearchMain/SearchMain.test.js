import React from 'react';
import { shallow } from 'enzyme';
import { mockEventData, mockCleanGeocodeData, mockCleanSearchData } from '../../mockData';
import { fetchAndCleanCategoryEventData } from '../../helper';
import { SearchMain, mapStateToProps, mapDispatchToProps } from './SearchMain';

describe('SearchMain', () => {
  let renderedComponent;
  let mockAddEvents;
  let mockRemoveSearch;
  let mockOnSearch;

  beforeEach(() => {
    window.fetch = jest.fn().mockImplementation(() => Promise.resolve({
      status: 200,
      json: () => Promise.resolve(mockEventData)
    }));
    mockAddEvents = jest.fn()
    mockRemoveSearch = jest.fn();
    mockOnSearch = jest.fn();
    renderedComponent = shallow(
      <SearchMain 
        addEvents={mockAddEvents}
        location={mockCleanGeocodeData}
        removeSearch={mockRemoveSearch}
        onSearch={mockOnSearch}
      />
    )
  })

  it('should match snapshot', () => {
    expect(renderedComponent).toMatchSnapshot();
  })

  it('should map to the store correctly', () => {
    const mockStore = {
      location: mockCleanGeocodeData,
    }
    const mapped = mapStateToProps(mockStore);

    expect(mapped).toEqual(mockStore);
    expect(mapped.location).toEqual(mockCleanGeocodeData);
  })

  it('should call the dispatch function when addEvents is called from MDTP', () => {
    const mockDispatch = jest.fn()
    const mapped = mapDispatchToProps(mockDispatch);
    mapped.addEvents();

    expect(mockDispatch).toHaveBeenCalled();
  });

  it('should call the dispatch function when removeSearch is called from MDTP', () => {
    const mockDispatch = jest.fn()
    const mapped = mapDispatchToProps(mockDispatch);
    mapped.removeSearch();

    expect(mockDispatch).toHaveBeenCalled();
  });

  describe('handleChange', () => {
    it('should update location state on input change', () => {
      const mockEvent = { target: { value: 'abc', name: 'location' }};
      const expected = 'abc';

      renderedComponent.instance().handleChange(mockEvent);

      renderedComponent.update();

      expect(renderedComponent.state().location).toEqual(expected);
    });

    it('should update eventSearch state on input change', () => {
      const mockEvent = { target: { value: 'soooo tired', name: 'eventSearch' }};
      const expected = 'soooo tired';

      renderedComponent.instance().handleChange(mockEvent);

      renderedComponent.update();

      expect(renderedComponent.state().eventSearch).toEqual(expected);
    })
  });

  describe('handleSubmit', () => {
    let mockEvent;

    beforeAll(() => {
      mockEvent = { preventDefault: jest.fn() }
    })

    it('should call fetch', () => {
      renderedComponent.instance().handleSubmit(mockEvent);

      expect(window.fetch).toHaveBeenCalled();
    })

    it('should remove what was in Search before', async () => {
      await renderedComponent.instance().handleSubmit(mockEvent);

      expect(mockRemoveSearch).toHaveBeenCalled();
    })

    it('should call addEvents with the expected params', async () => {
      const mockSearchedEvents = mockCleanSearchData;
      
      await renderedComponent.instance().handleSubmit(mockEvent);

      expect(mockAddEvents).toHaveBeenCalledWith(mockSearchedEvents, 'search');
    })

    it('should call onSearch to reroute page', async () => {
      await renderedComponent.instance().handleSubmit(mockEvent);

      expect(mockOnSearch).toHaveBeenCalled();
    })
  });  
})