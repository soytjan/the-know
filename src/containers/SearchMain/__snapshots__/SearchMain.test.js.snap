// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchMain should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchMain
    addEvents={[Function]}
    location={
        Object {
            "address": "Denver, CO, USA",
            "coordinates": Object {
              "lat": 39.7392358,
              "lng": -104.990251,
            },
          }
    }
    onSearch={[Function]}
    removeSearch={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form
        onSubmit={[Function]}
>
        <input
                className="input-search"
                name="eventSearch"
                onChange={[Function]}
                placeholder="Find events in your area"
                type="text"
                value=""
        />
        <input
                className="input-location"
                name="location"
                onChange={[Function]}
                placeholder="Where?"
                type="text"
                value="Denver, CO, USA"
        />
        <button>
                SEARCH
        </button>
</form>,
      "className": "SearchMain",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="input-search"
            name="eventSearch"
            onChange={[Function]}
            placeholder="Find events in your area"
            type="text"
            value=""
/>,
          <input
            className="input-location"
            name="location"
            onChange={[Function]}
            placeholder="Where?"
            type="text"
            value="Denver, CO, USA"
/>,
          <button>
            SEARCH
</button>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "input-search",
            "name": "eventSearch",
            "onChange": [Function],
            "placeholder": "Find events in your area",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "input-location",
            "name": "location",
            "onChange": [Function],
            "placeholder": "Where?",
            "type": "text",
            "value": "Denver, CO, USA",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "SEARCH",
          },
          "ref": null,
          "rendered": "SEARCH",
          "type": "button",
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form
          onSubmit={[Function]}
>
          <input
                    className="input-search"
                    name="eventSearch"
                    onChange={[Function]}
                    placeholder="Find events in your area"
                    type="text"
                    value=""
          />
          <input
                    className="input-location"
                    name="location"
                    onChange={[Function]}
                    placeholder="Where?"
                    type="text"
                    value="Denver, CO, USA"
          />
          <button>
                    SEARCH
          </button>
</form>,
        "className": "SearchMain",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <input
              className="input-search"
              name="eventSearch"
              onChange={[Function]}
              placeholder="Find events in your area"
              type="text"
              value=""
/>,
            <input
              className="input-location"
              name="location"
              onChange={[Function]}
              placeholder="Where?"
              type="text"
              value="Denver, CO, USA"
/>,
            <button>
              SEARCH
</button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "input-search",
              "name": "eventSearch",
              "onChange": [Function],
              "placeholder": "Find events in your area",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "input-location",
              "name": "location",
              "onChange": [Function],
              "placeholder": "Where?",
              "type": "text",
              "value": "Denver, CO, USA",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "SEARCH",
            },
            "ref": null,
            "rendered": "SEARCH",
            "type": "button",
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
